Add a simple particle system

Dreauw opened the pull request:
-- Team 2 --

We needed a simple [particle system](https://en.wikipedia.org/wiki/Particle_system) in our project.
As it seems to have its place in a game framework, I made this PR, but feel free to reject it if you think that it's too much.

So, I will briefly explain my implementation : 
At first, I wanted every particles to be a game entity, but it was killing performances and there were issues like "superposition" (As it is right now, entites are drawn in the same order that they are added. But particles were always added after everything, so they were drawn above everything).
So for all these reasons, I decided to make only one game entity (ParticleEmitter) that will stores (with a list of "Particle") and draw every particle.
For the behavior of the particles, I decided to adapt a decorator pattern because behaviors needs to be combined and it makes it really easy for anybody to add a new behavior.
A single instance of behavior can be applied to as many particle as you want. I did it that way to avoid having a lot of instances created (and therefore, a lot of garbage in memory) per particle (as we could possibly have a lot of particles simultaneously).


As it is more a graphical kind of thing, I will put an example here to show what my implementation is capable of.
For a game running at 60FPS, this example : 
```
emitter = new ParticleEmitter();
universe.addGameEntity(emitter);
[...]	
ParticleBehavior behavior = new DefaultParticleBehavior();
behavior = new DyingParticleBehavior(behavior, 200, true);
behavior = new MovingParticleBehavior(behavior, 5, 0, -Math.PI);
behavior = new DelayedParticleBehavior(behavior, 1);
behavior = new GravityParticleBehavior(behavior, 0.15, 0.15);
		
Rectangle rect = new Rectangle(450, 100, 4, 7);
emitter.emit(Color.RED, rect, 200, behavior);
```

Will give that effect : http://i.imgur.com/Rkh7aX4.gif

DamienCassou commented on src/main/java/gameframework/particles/Particle.java:
```diff
@@ -0,0 +1,167 @@
+		return x;
+	}
+	/**
+	 * @param x the x to set
```
useless comment

DamienCassou commented on src/main/java/gameframework/particles/Particle.java:
```diff
@@ -0,0 +1,167 @@
+	
+	/**
+	 * Constructor
+	 * @param id Id of the particle
```
which Id should a user pass? Why can't the class take care of generating IDs itself?

DamienCassou commented on src/main/java/gameframework/particles/Particle.java:
```diff
@@ -0,0 +1,167 @@
+	protected int timeAlive;
+	protected Color color;
+	protected double randomness;
+	protected int id;
```
why is the ID useful?

DamienCassou commented on src/main/java/gameframework/particles/Particle.java:
```diff
@@ -0,0 +1,167 @@
+
+/**
+ * Stores every useful information about a single particle
+ * like its position on the screen, its color...
```
say something about related classes such as the emitter

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+ */
+public class ParticleEmitter implements GameEntity {
+	/* List containing every "alive" particles that will be updated */
+	protected List<Particle> particles;
```
why not `aliveParticules`?

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+	 * @param particle Particle to update
+	 */
+	protected void update(Particle particle) {
+		particle.setTimeAlive(particle.getTimeAlive() + 1);
```
you could simplify with `particle.incrementTimeAlive()`

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+	protected void update(Particle particle) {
+		particle.setTimeAlive(particle.getTimeAlive() + 1);
+		
+		if (particle.getBehavior().isDead(particle)) {
```
why not simply `particle.isDead()`?

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+		particle.setTimeAlive(particle.getTimeAlive() + 1);
+		
+		if (particle.getBehavior().isDead(particle)) {
+			deadParticles.add(particle);
```
shouldn't you remove the particle from the alive particles as well?

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+			deadParticles.add(particle);
+		}
+		
+		if (particle.getBehavior().isDrawn(particle)) {
```
`particle.isDrawn()`

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+		}
+		
+		if (particle.getBehavior().isDrawn(particle)) {
+			particle.getBehavior().update(particle);
```
`particle.update()`

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+	 * Remove every dead particles contained in deadParticles
+	 */
+	protected void removeDeadParticles() {
+		for (Particle p : deadParticles) {
```
use `List.removeAll()`

DamienCassou commented on src/main/java/gameframework/particles/ParticleEmitter.java:
```diff
@@ -0,0 +1,106 @@
+	 * @param g The Graphics
+	 */
+	public void drawParticle(Particle p, Graphics g) {
+		if (p.getBehavior().isDrawn(p)) {
```
You can realize that this implementation has a problem because it doesn't use any instance variable. That's a code smell. It indicates that the code is probably not in the right place. You could replace it with `p.draw(g)` for example, but you probably don't need this method at all.

Please also avoid `p` and `g` as variable names

DamienCassou commented on src/main/java/gameframework/particles/behaviors/DefaultParticleBehavior.java:
```diff
@@ -0,0 +1,41 @@
+
+	/**
+	 * By default, the particle cant die
+	 * @see nidhogglike.particles.behaviors.ParticleBehavior#isDead(nidhogglike.particles.Particle)
```
what is `nidhogglike`?

DamienCassou commented on src/main/java/gameframework/particles/behaviors/DelayedParticleBehavior.java:
```diff
@@ -0,0 +1,31 @@
+import gameframework.particles.Particle;
+
+/**
+ * This behavior will adds a delay between every particle
```
english typo

DamienCassou replied:
what do you mean by "a delay between particle"?

DamienCassou commented on src/main/java/gameframework/particles/behaviors/GravityParticleBehavior.java:
```diff
@@ -0,0 +1,38 @@
+import gameframework.particles.Particle;
+
+/**
+ * This behavior apply a gravitational force to every particle
```
english typo

Dreauw commented:
> which Id should a user pass?

A "Particle" should be instanciated exclusively by the "ParticleEmitter".
The user will never use this class directly so he doesn't have to worry about this id. (But I have to admit that this could be misleading. So to avoid errors, I changed the constructor to "protected", like that only the ParticleEmitter will be able to instanciate a Particle as the two classes are in the same package)
I added comments to inform the user that "Particle" is used by the ParticleEmitter and shouldn't be used

> Why can't the class take care of generating IDs itself?

This class can't take care of generating this ID 'cause the id is relative to the "group" of the particle
And we only have this information in the ParticleEmitter since it's this class that will create the "group" of particle (when we call ParticleEmitter#emit)
I added comments about that

> why is the ID useful? How is it different from the Particular instance itself?

The id can be used as a way to determine the order of the particles inside of a group of particles emitted after a call of ParticleEmitter#emit
For example, "DelayedParticleBehavior" is using the id to determine when to draw a certain particle. Like that, the particle with the id 0 will be drawn first, then, after a delay, the particle with the id 1 is showed and so on...


> You can realize that this implementation has a problem because it doesn't use any instance variable. That's a code smell. It indicates that the code is probably not in the right place. You could replace it with p.draw(g) for example, but you probably don't need this method at all.

This code is here because it's the code responsible for the drawing of the particles. Like that, the user can override this method if he wants the particle to be, for example, circles or images.
If I make a method "draw" in Particle, then the user will not be able to easily redefines how the particles are drawn 'cause "Particle" is instantiated in "ParticleEmitter". So he would have to override ParticleEmitter and Particle.
I don't really know where to put this code otherwise.



Apart from that, I did all the changes that you proposed.

DamienCassou commented:
You refer to a "group" of particles and to "id" that only make sense within this group. I have the impression that your code misses this "group" abstraction. That way, the particles don't have an ID and can be reused without the emitter (you never know what other code might want to reuse) and only a particle group maintain particle IDs. What do you think?

DamienCassou commented:
I would like to merge that. Do you want to take care of the feedback?

Dreauw commented:
Done.

DamienCassou commented:
Thank you very much
